processor/tailsamplingprocessor/internal/sampling/composite_test.go

```diff
@@ -0,0 +1,250 @@
+	"go.uber.org/zap"
+)
+
+type FakeTimeProvider struct {
+	second int64
+}
+
+func (f FakeTimeProvider) getCurSecond() int64 {
+	return f.second
+}
+
+var traceID = pdata.NewTraceID([16]byte{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x52, 0x96, 0x9A, 0x89, 0x55, 0x57, 0x1A, 0x3F})
+
+func createTrace() *TraceData {
+	trace := &TraceData{SpanCount: 1}
+	return trace
+}
+
+func TestCompositeEvaluatorNotSampled(t *testing.T) {
+
+	// Create 2 policies which do not match any trace
+	n1 := NewNumericAttributeFilter(zap.NewNop(), "tag", 0, 100)
+	n2 := NewNumericAttributeFilter(zap.NewNop(), "tag", 200, 300)
+	c := NewComposite(zap.NewNop(), 1000, []SubPolicyEvalParams{{n1, 100}, {n2, 100}}, FakeTimeProvider{})
+
+	trace := createTrace()
+
+	decision, err := c.Evaluate(traceID, trace)
+	if err != nil {
+		t.Fatalf("Failed to evaluate composite policy: %v", err)
+	}
+
+	// None of the numeric filters should match since input trace data does not contain
+	// the "tag", so the decision should be NotSampled.
+	expected := NotSampled
+	if decision != expected {
+		t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+	}
+}
+
+func TestCompositeEvaluatorSampled(t *testing.T) {
+
+	// Create 2 subpolicies. First results in 100% NotSampled, the second in 100% Sampled.
+	n1 := NewNumericAttributeFilter(zap.NewNop(), "tag", 0, 100)
+	n2 := NewAlwaysSample(zap.NewNop())
+	c := NewComposite(zap.NewNop(), 1000, []SubPolicyEvalParams{{n1, 100}, {n2, 100}}, FakeTimeProvider{})
+
+	trace := createTrace()
+
+	decision, err := c.Evaluate(traceID, trace)
+	if err != nil {
+		t.Fatalf("Failed to evaluate composite policy: %v", err)
+	}
+
+	// The second policy is AlwaysSample, so the decision should be Sampled.
+	expected := Sampled
+	if decision != expected {
+		t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+	}
+}
+
+func TestCompositeEvaluatorSampled_AlwaysSampled(t *testing.T) {
+
+	// Create 2 subpolicies. First results in 100% NotSampled, the second in 100% Sampled.
+	n1 := NewNumericAttributeFilter(zap.NewNop(), "tag", 0, 100)
+	n2 := NewAlwaysSample(zap.NewNop())
+	c := NewComposite(zap.NewNop(), 10, []SubPolicyEvalParams{{n1, 20}, {n2, 20}}, FakeTimeProvider{})
+
+	for i := 1; i <= 3; i++ {
+		trace := createTrace()
+
+		decision, err := c.Evaluate(traceID, trace)
+		if err != nil {
+			t.Fatalf("Failed to evaluate composite policy: %v", err)
+		}
+
+		// The second policy is AlwaysSample, so the decision should be Sampled.
+		expected := Sampled
+		if decision != expected {
+			t.Fatalf("Incorrect decision by composite policy evaluator: expected %v, actual %v", expected, decision)
+		}
+	}
+}
+
+func TestCompositeEvaluatorThrottling(t *testing.T) {
+
+	// Create only one subpolicy, with 100% Sampled policy.
+	n1 := NewAlwaysSample(zap.NewNop())
+	timeProvider := &FakeTimeProvider{second: 0}
+	const totalSPS = 100
```

Comment:
You can probably set a lower SPS, requiring your test to do fewer things.

Reply:
The same applies to further places in this PR.

Reply:
done